Add support for OpenID Connect (OIDC)

timhallmann opened the pull request.

FlorianRhiem reviewed (changes requested):
Hey Tim, thank you for the PR!

I've gone through the code and added a few comments, though nothing major. If possible, you might also want to add missing translations, currently the following translatable strings are reported:

```diff
$ diff --color=always extracted_messages_current_de_without_header.po extracted_messages_generated_de_without_header.po
2312a2313,2316
> #, fuzzy
> msgid "OIDC"
> msgstr "Themen"
> 
2993a2998,3001
> #, fuzzy, python-format
> msgid "Sign in with %(oidc_name)s"
> msgstr "Anmelden mit %(component_name)s"
> 
7252a7261,7264
> 
> #, fuzzy
> msgid "Login failed. Please contact an administrator."
> msgstr "Die Nutzer-ID war ungültig. Bitte kontaktieren Sie einen Administrator."
```

mypy also reports some small issues, mostly overuse of `type: ignore`:

```
$ MYPYPATH=./stubs python3 -m mypy --strict --ignore-missing-imports --follow-imports=silent sampledb/
sampledb/logic/oidc.py:126: error: Unused "type: ignore" comment  [unused-ignore]
sampledb/logic/oidc.py:150: error: Returning Any from function declared to return "str"  [no-any-return]
sampledb/logic/oidc.py:221: error: Returning Any from function declared to return "str"  [no-any-return]
sampledb/logic/oidc.py:238: error: Unused "type: ignore" comment  [unused-ignore]
sampledb/logic/oidc.py:239: error: Unused "type: ignore" comment  [unused-ignore]
sampledb/logic/oidc.py:241: error: Unused "type: ignore" comment  [unused-ignore]
sampledb/logic/oidc.py:248: error: Module "sampledb.logic.users" does not explicitly export attribute "users"  [attr-defined]
sampledb/logic/oidc.py:302: error: Unused "type: ignore" comment  [unused-ignore]
```

Would it be possible to add tests, e.g. with a minimal mock implementation of OIDC?

timhallmann commented on docs/user_guide/users.rst:
```diff
@@ -8,9 +8,31 @@ Users
+
+4. Invitation-based Authentication
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+* For guests or users at facilities where the previous two authentication methods are not available:
```
Indeed, I forgot to change that. I wrote the docs a while ago.

timhallmann commented on docs/user_guide/users.rst:
```diff
@@ -84,6 +106,7 @@ Deactivated Users
 
 Users can also be deactivated. These users will be unable to sign in to their account or use the API until they have been reactivated by an administrator. As they will be unable to access their own data, this should only be used if marking a user as readonly will not suffice.
 
+.. _users_federation:
```
I intended to do so but removed the link. I removed the anchor as well.

timhallmann commented on docs/administrator_guide/configuration.rst:
```diff
@@ -67,6 +67,35 @@ LDAP
+   * - SAMPLEDB_OIDC_SCOPES
+     - The scopes to request, defaults to ``openid profile email``. Must include the ``email``, ``email_verified`` and ``name`` properties.
+   * - SAMPLEDB_OIDC_ROLES
+     - If set, reads roles using this attribute path. See the following section on roles.
```
Yes. I've removed the sentence, `oidc.rst` is already linked via the following paragraph.

timhallmann commented on docs/administrator_guide/configuration.rst:
```diff
@@ -67,6 +67,35 @@ LDAP
+     - The scopes to request, defaults to ``openid profile email``. Must include the ``email``, ``email_verified`` and ``name`` properties.
+   * - SAMPLEDB_OIDC_ROLES
+     - If set, reads roles using this attribute path. See the following section on roles.
+   * - SAMPLEDB_OIDC_ONLY
+     - If set, other login methods and invitations will be disabled. This does not affect API keys.
```
It's a bit of a workaround to make `/users/me/sign_in` redirect to the OIDC provider instead of displaying the login page. It's not strictly necessary, but it does eliminate unnecessary clicks and the need to choose an authentication method.

Although I just realized that after switching to the check in `_sign_in_impl`, it's no longer necessary to hide the other authentication methods in `sampledb/frontend/templates/sign_in.html`.

FlorianRhiem replied:
I see the use of having a way to check that OIDC is the only enabled authentication method, I only thought that an additional configuration variable which only overrides existing ones feels like an odd way to do it. In a "sane" configuration where `OIDC_ONLY` doesn't directly conflict with and override other variables, I would assume that LDAP isn't configured, both FIDO2 passkey authentication and federated logins aren't enabled and - the only thing different to the default - invitations are disabled.

So, unless I'm missing something, setting `OIDC_ONLY` could be replaced by setting `DISABLE_USER_INVITATIONS`. Checking for `OIDC_ONLY`, e.g. to implement the mechanisms you mentioned, could then be replaced with checking via function, similar to `is_oidc_configured` that checks that `DISABLE_USER_INVITATIONS` is true, `SAMPLEDB_ENABLE_FEDERATED_LOGIN` is false, `SAMPLEDB_ENABLE_FIDO2_PASSKEY_AUTHENTICATION` is false and either that `is_ldap_configured()` is false?

timhallmann replied:
We could do that if we add a variable to check if the login using email / username and password is disabled as well. Unless we wish to overload `DISABLE_USER_INVITATIONS`, which I wouldn't suggest.

Though I'd prefer a more explicit approach where the redirect is intentionally enabled, rather than being a side effect of disabling everything except OIDC.

FlorianRhiem replied:
Ah, right, I didn't think of bot accounts and existing invitations. Right, in that case, rather than adding new variables to disable these, let's just keep `OIDC_ONLY` as it is here.

timhallmann commented on sampledb/logic/oidc.py:
```diff
@@ -0,0 +1,323 @@
+    """
+    Store the login state.
+
+    The url is used to communicate the redirect_uri as well as the
```
That comment was outdated, I've removed it. Only the URL of the `?next` query arg is stored.

timhallmann commented on sampledb/logic/oidc.py:
```diff
@@ -0,0 +1,323 @@
+    if (
+        'code_challenge_methods_supported' in client.provider_config.__dict__ and
+        'S256' in client.provider_config.code_challenge_methods_supported  # type: ignore
+    ):
```
No idea. Fixed.

timhallmann commented:
I added translations using the following commands I found in the `.gitlab-ci.yml`. Looks okay though, similar to prior changes of `extracted_messages.po`.

    pybabel extract --no-location -F babel.cfg -k lazy_gettext -o sampledb/messages.pot sampledb
    pybabel update -i sampledb/messages.pot -d sampledb/translations -D extracted_messages

The mypy issues are mostly a result of differences between mypy and pyright. While mypy treats libraries that don't have a `py.typed` as untyped, pyright does treat them as typed, so I didn't notice py_simple_openid_connect lacking a `py.typed` file. I've submitted a PR and patched this in my fork.

I've added logic tests. Tests for the frontend seem not that worthwhile to me but I could test the behaviour of `SAMPLEDB_OIDC_ONLY` if you consider that necessary.

timhallmann commented:
I added a small change to update the users name and email during a regular login if the ID Token contains all necessary attributes, and also shuffled some tests around.

FlorianRhiem commented on sampledb/frontend/users/authentication.py:
```diff
@@ -270,3 +283,38 @@ def shared_device_state() -> FlaskResponseT:
+
+@frontend.route('/users/me/oidc/start')
+def oidc_start() -> FlaskResponseT:
+    url = oidc.start_authentication(
```
The routes `/users/me/oidc/start` and `/users/me/oidc/callback` should check whether OIDC is configured (and return 404 otherwise) and whether the user is already signed in (and redirect them away if so).

timhallmann replied:
So, without explicit error handling, we'd get a KeyError once we fetch the client and thus always return 500.

A redirect seems fine as well but I'm not so sure about choosing either a redirect or an error depending on the sign in status, especially if we switch the return code to 404.

Now that I've looked for it, federated login currently flashes a "Federated login is disabled" message and redirects to the index (regardless of sign in status), which makes sense as well and which I could add later this day.

FlorianRhiem replied:
Sounds good!

timhallmann replied:
Done.

FlorianRhiem replied:
The check for already being signed in is still missing. Instead, going to the route seems to just re-do the sign-in process.

timhallmann replied:
Done.

FlorianRhiem commented on sampledb/logic/authentication.py:
```diff
@@ -338,6 +338,49 @@ def add_federated_login_authentication(federated_identity: FederatedIdentity) ->
+    Get all authentication methods of type OIDC with the given issuer and
+    subject identifier.
+
+    :param user_id: the ID of an existing user
```
```suggestion
```

timhallmann replied:
Done.

FlorianRhiem commented on sampledb/frontend/users/preferences.py:
```diff
@@ -156,6 +156,10 @@ def _handle_account_information_forms(
 def _handle_authentication_methods_forms(
         template_kwargs: typing.Dict[str, typing.Any]
 ) -> typing.Optional[FlaskResponseT]:
+    if logic.oidc.is_oidc_only_auth_method():
```
This function defines a few template parameters which are mostly used for the forms themselves, however `authentications` and `api_access_tokens` should be defined even if these forms are now disabled in this case, otherwise users cannot see the authentication methods (i.e. OIDC) and API access tokens they have. While the API access tokens should be displayed either way, it could be reasonable to simply hide the whole authentication method table in the preferences template if `OIDC_ONLY` is set. In that case, `authentications` would not be necessary.

With `authentication_form` and `options` not being set, a check for `authentication_form` should be placed around the `addAuthenticationMethodModal` in preferences.html and a check for `options` should be placed around the line`{% do set_template_value("fido2_options", dict(options)) %}`.

timhallmann replied:
I moved the check below the update of the `template_kwargs`. There's already a check in `preferences.html` to hide the button to open the modal if `OIDC_ONLY` is set.

I'm in favour of keeping it that way as it minimizes differences based on configuration variables and still has a server-side check, but I'm also fine with your approach.

FlorianRhiem reviewed (changes requested):
When testing this with an existing user that matches the user from the identity provider by email, but not by user name, the user name is only updated after the second login. The first finds the match and creates the authentication method, the second sees the existing match and updates the user name. `_update_user` should be called during the first login, after matching the existing user.

timhallmann commented on sampledb/logic/oidc.py:
```diff
@@ -0,0 +1,341 @@
+    :param id_token_hint: The ID token returned in the authentication.
+    :param post_logout_redirect_uri: The URL to which the OP should redirect
+        the user.
+    :return: A tuple of the user, the ID token, and the next URL.
```
Fixed.

timhallmann commented on sampledb/logic/oidc.py:
```diff
@@ -0,0 +1,341 @@
+        id_token_hint=id_token_hint,
+        client_id=client.client_auth.client_id,
+        post_logout_redirect_uri=post_logout_redirect_uri,
+    ))
```
Done.

timhallmann commented on sampledb/logic/oidc.py:
```diff
@@ -0,0 +1,341 @@
+    IdToken, RpInitiatedLogoutRequest, TokenErrorResponse,
+    TokenSuccessResponse, UserinfoErrorResponse
+)
+
```
Done.

timhallmann commented on sampledb/logic/oidc.py:
```diff
@@ -0,0 +1,341 @@
+
+    id_token = client.decode_id_token(
+        token_response.id_token,
+        nonce=current_url.args['nonce'],
```
I've also added the `OIDC_DISABLE_NONCE` config variable.

As I understand it, a conforming provider must support `nonce` values, but compatibility seems a bit more useful here than strict adherence to (my understanding of) the spec or security.

FlorianRhiem commented on sampledb/logic/oidc.py:
```diff
@@ -0,0 +1,340 @@
+    function. The next URL is the address the user tried to visit before being
+    asked to authenticate.
+
+    :param url: The request url.
```
```suggestion
    :param url: The request url.
    :param token: The JSON token originally generated by start_authentication
```

timhallmann replied:
Done. I've removed the mention of JSON, I'd like to keep the format internal to `oidc.py`. For everyone else, it's just an opaque string.

FlorianRhiem commented on sampledb/logic/oidc.py:
```diff
@@ -0,0 +1,340 @@
+import flask
+import pydantic
+from furl import furl
+from simple_openid_connect import pkce
```
```suggestion
```

FlorianRhiem commented on sampledb/config.py:
```diff
@@ -123,6 +130,7 @@ def parse_configuration_values() -> None:
         'DISABLE_TOPICS',
         'ENABLE_FEDERATED_LOGIN',
         'ENABLE_FEDERATED_LOGIN_CREATE_NEW_USER',
+        'OIDC_ONLY',
```
```suggestion
        'OIDC_ONLY',
        'OIDC_DISABLE_NONCE',
```

Hopefully this'll be the final small change :)

timhallmann replied:
... because all other final changes will require a major rewrite? ;)

Done.

FlorianRhiem commented:
Thank you again for the PR!

Could you write a post to the [mailing list](https://lists.fz-juelich.de/mailman/listinfo/sampledb) at sampledb@fz-juelich.de introducing the new feature? :)
